File: Diffie-Hellman.py
#!/usr/bin/python

#Diffie-Hellman Key Cryptographic key exchange in python
#Before using this program, ensure that you have exchanged
#The prime modulus and base with your partner
import time


def isInt(n):
  #Function for checking if secret is an integer
    return type(1) == type(n)


def isPrime(n):
  #Function for checking if mod and base are prime
    n = abs(int(n))
    if n < 2:
        return False
    if n == 2:
        return True
    if not n & 1:
        return False
    for x in range(3, int(n**0.5)+1, 2):
        if n % x == 0:
            return False
    return True

#introduction
print ""
print "*********************************"
print "*  Diffie-Hellman Key Exchange  *"
print "*  ---------------------------  *"
print "*    Authored by Jack Renshaw   *"
print "*********************************"
print ""

base = input("Enter your shared prime base: ") #Request user for (shared) prime base
modulus = input("Enter your shared prime modulus: ") #Request user for (shared) prime modulus 
secret = input("Please enter your secret key: ")

#Run the secret, mod and base through the functions
if not isPrime(base):
    print "Your base is not prime..."
    exit()
if not isPrime(modulus):
    print "Your modulus is not prime..."
    exit()
if not isInt(secret):
    print "Your secret is not an integer..."
    exit()

print "calculating result..." #Really stupid and gimmicky
time.sleep(2)

print "Share this result with your partner: ", base**secret % modulus #Calculate the result
print ""
uresult = input("Please enter your partners result: ") #Request user for partners result
print "calculating shared key..." #Stupid and gimmicky
time.sleep(2)
print "Your shared secret key is: ", uresult**secret % modulus #Calculate the shared secret from partners function result and your secret key
print ""

print "Congralutations, you can now communicate in secret" #finitto
print ""

File: Diffie-Hellman.cpp
#include <iostream> //input output header
#include <cmath> //math header

int PrimeChecker(int n) // Function to check if number is prime
{
    for(int i=2; i < n; i++)
    {
       if(n%i!=0)
          return 1;
       else 
          return 0;
    }  
}

int main() // main function
{

	using namespace std;
	int base, secret, modulus, uresult, b1, b2, b2s, r2s; //declare variables

	cout << "\n"; //introduction
	cout << "*********************************\n";
	cout << "*  Diffie-Hellman Key Exchange  *\n";
	cout << "*  ---------------------------  *\n";
	cout << "*    Authored by Jack Renshaw   *\n";
	cout << "*********************************\n";
	cout << "\n";

	cout << "Enter your shared prime base: "; //Request prime base from user
	cin >> base; // input base
	cout << "Enter your shared prime modulus: "; //Request prime modulus from user
	cin >> modulus; // input modulus
	cout << "Enter your secret key: "; //Request secret number from user
	cin >> secret; // input secret, must already be int

	b1 = PrimeChecker(base); //verify that base is prime
	b2 = PrimeChecker(modulus); // verify that modulus is prime

	if(b1 != 1){
		cout << "Your base is not prime...\n"; //return error
	}
	if(b2 != 1){
		cout << "Your modulus is not prime...\n"; //return error
	}

	b2s = pow(base, secret); // calculate the base to the power of secret

	cout << "Share this result with your partner: " << b2s % modulus << endl << endl; // Return result
	cout << "Enter your partners result: "; // Request partners result
	cin >> uresult; // input result

	r2s = pow(uresult, secret); // calculate the parter result to the power of secret
	cout << "Your shared secret is: " << r2s % modulus << endl << endl; // Return shared key

	cout << "Congratulations, you can now communicate in secret" << endl << endl; // Finnito

	return 0;

}